whitepapers/0007_killswitch.md

```diff
@@ -0,0 +1,87 @@
+* `votes`: a `list` of security guardian votes as signers + bools (true = up, false = down)
+
+For each send/receive capability for a token/nft bridge contract, it must implement a check against votes to ensure a majority up vote to enable/disable transactions.
+
+To initiate a vote, each security guardian must retrieve the current `nonce` and send a `signed message` that includes the `nonce` and their `vote`.  Once a `vote` is placed, it increments the nonce by 1.  Any implementation must check the `signed message` and make sure it matches the existing `nonce`  and is a valid signer before registering a vote.
+
+Each smart contract must implement one function:
+
+* `shutdown`: a function that accepts `signed_message` which includes `nonce` and `vote` from a given signer.  This method will validate the signer ID is in `signers` and the `nonce` is valid (to prevent replay attacks) before adding the `vote` (true = down, false = up) for that signer and incrementing the nonce.
+
+### Pros
+
+* Eliminates the burden of binding shutdown actions from a specific wallet
+* Normal governance can be used to update signers as the security gaurdians change
+
+### Cons
+
+* Will require some CLI tooling or manual steps for a signer to obtain the current nonce before signing a message to ensure speed
+* If for some reason the nonce rolls over and becomes non-unique, it could be succeptable to replay attacks that the nonce aims to solve for
+* Will require each signer to wait until the next block before signing
+
+
+## Detailed Design (Option 2)
+
+Each smart contract must implement the following 3 storage components:
+
+* `wallets`: a `list` of wallets for which we will receive signed votes from
+* `signers`: a `list` keys of security guardians within the security counsel who are authorized to vote
+* `votes`: a `list` of security guardian votes as signers + bools (true = up, false = down)
+
+For each send/receive capability for a token/nft bridge contract, it must implement a check against votes to ensure a majority up vote to enable/disable transactions.
+
+Each smart contract must implment one function:
+
+* `shutdown`: a function that accepts `signed_message` which includes `vote` from a given signer.  This method will validate the `signer` and the `wallet` are in `signers` and `wallets` before adding the `vote` (true = down, false = up) for that signer.
+
+### Pros
+
+* Less complexity and thus less potential for abuse when compared to option 1
+* Normal governance can be used to update signers as the security gaurdians change
+
+### Cons
+
+* Security Guardians would be limited from sending a `vote` from a single wallet.  Alternatively, if an implementation didn't bind wallets and signers 1:1, a single guardian could send `signed message` on another security guardians behalf if that guardian couldn't use their wallet.
+
+## Detailed Design (Option N)
+
+If you have other proposals, please add them here or suggest them via the PR.
```

Comment:
Here’s an alternative idea: instead of a list of signers and booleans, have a list of token accounts, that are controlled by the signers. Put some tokens in it (it doesn’t really matter which ones or how many). The contract would check that a majority of those accounts have a positive balance. To trigger the kill switch, signers temporarily move the funds to a different account.

### Pros

* Doesn’t require any special tooling, signers can just use their everyday wallet, which means it should be very quick. (I.e. no figuring out last moment that the CLI utility is broken and needs to be rebuilt from source, or figuring out what the right syntax was for referencing a particular derivation path on the Ledger, etc.)
* Can easily check the status on block explorers without custom tooling.

### Cons

* Infeasible on Solana because all contracts would need all of these accounts as input and Wormhole already needs a lot of account inputs.
* Not possible for a 2/3 majority of the guardians to re-enable the network after the kill switch is triggered, it can only be undone by the same people who triggered it. Although I suppose this could be addressed with a contract update if needed.

I don’t think this is a feasible alternative, but maybe some iteration of it can be helpful.

Reply:
Reverted to quorum on this for now, keeps the security assumptions a bit cleaner, but it's a trade off on some speed.